File: src/web-ui/src/infrastructure/api/service-api/ReviewPlatformAPI.ts
import { api } from './ApiClient';
import { createTauriCommandError } from '../errors/TauriCommandError';
import { createLogger } from '@/shared/utils/logger';

const log = createLogger('ReviewPlatformAPI');

export type ReviewPlatformKind = 'github' | 'gitlab' | 'gitcode' | 'gitee' | 'unknown';
export type ReviewAuthState = 'not_connected' | 'not_required' | 'connected' | 'expired' | 'error' | 'unsupported';
export type ReviewAuthSource = 'gh_cli' | 'env' | 'stored' | 'none' | 'unsupported';
export type ReviewAuthChallengeState = 'missing' | 'invalid' | 'insufficient_scope';
export type ReviewItemState = 'open' | 'merged' | 'closed' | 'draft';
export type ReviewPlatformListState = 'all' | ReviewItemState;
export type ReviewDecision = 'approved' | 'changes_requested' | 'commented' | 'pending';
export type ReviewFileStatus = 'added' | 'modified' | 'deleted' | 'renamed';
export type ReviewPlatformDetailSection = 'overview' | 'ci' | 'files' | 'commits' | 'reviews';
export type ReviewEvidenceCompleteness = 'complete' | 'partial';

export interface ReviewPlatformAccount {
  id: string;
  platform: ReviewPlatformKind;
  label: string;
  username?: string | null;
  host: string;
  authState: ReviewAuthState;
  authSource: ReviewAuthSource;
  scopes: string[];
  message?: string | null;
}

export interface ReviewPlatformRepositoryRef {
  providerId: string;
  platform: ReviewPlatformKind;
  host: string;
  owner: string;
  name: string;
  projectPath: string;
  defaultBranch: string;
  workspacePath?: string | null;
  webUrl: string;
}

export interface ReviewPlatformAuthChallenge {
  platform: ReviewPlatformKind;
  host: string;
  remoteId: string;
  projectPath: string;
  state: ReviewAuthChallengeState;
  message: string;
  requiredScopes: string[];
}

export interface ReviewPlatformRemote {
  id: string;
  name: string;
  url: string;
  platform: ReviewPlatformKind;
  host: string;
  owner: string;
  repositoryName: string;
  projectPath: string;
  webUrl: string;
  supported: boolean;
  authState: ReviewAuthState;
  authSource: ReviewAuthSource;
  message?: string | null;
}

export interface ReviewChecks {
  total: number;
  passed: number;
  failed: number;
  pending: number;
}

export interface ReviewPlatformCiItem {
  id: string;
  name: string;
  status: string;
  conclusion?: string | null;
  detail?: string | null;
  stage?: string | null;
  webUrl?: string | null;
  log?: string | null;
  logTruncated: boolean;
  startedAt?: string | null;
  finishedAt?: string | null;
}

export interface ReviewPlatformPullRequest {
  id: string;
  providerId?: string | null;
  number: number;
  title: string;
  state: ReviewItemState;
  author: string;
  sourceBranch: string;
  targetBranch: string;
  baseRevision?: string | null;
  headRevision?: string | null;
  updatedAt: string;
  webUrl: string;
  additions: number;
  deletions: number;
  lineStatsKnown?: boolean;
  changedFiles: number;
  /** Missing on older backends; only an explicit false means the count is unknown. */
  changedFileCountKnown?: boolean;
  comments: number;
  reviewDecision: ReviewDecision;
  checks: ReviewChecks;
}

export interface ReviewPlatformFile {
  path: string;
  oldPath?: string | null;
  status: ReviewFileStatus;
  additions: number;
  deletions: number;
  patch?: string | null;
}

export interface ReviewPlatformReviewTargetFile {
  path: string;
  oldPath?: string | null;
  status: ReviewFileStatus;
  additions: number;
  deletions: number;
  diffAvailable: boolean;
}

export interface ReviewPlatformPullRequestReviewTarget {
  pullRequest: ReviewPlatformPullRequest;
  files: ReviewPlatformReviewTargetFile[];
  omittedFileCount: number;
  limitations: string[];
}

export interface ReviewPlatformIssueComment {
  id: string;
  webUrl?: string | null;
  author?: string | null;
  body: string;
  createdAt?: string | null;
  updatedAt?: string | null;
}

export interface ReviewPlatformIssueEvidence {
  platform: ReviewPlatformKind;
  host: string;
  projectPath: string;
  issueId: string;
  webUrl: string;
  title: string;
  body: string;
  state: string;
  author?: string | null;
  labels: string[];
  createdAt?: string | null;
  updatedAt?: string | null;
  comments: ReviewPlatformIssueComment[];
  fingerprint: string;
  completeness: ReviewEvidenceCompleteness;
  limitations: string[];
  hasMoreComments: boolean;
  nextCursor?: string | null;
}

export interface ReviewPlatformCommit {
  hash: string;
  shortHash: string;
  title: string;
  author: string;
  committedAt: string;
}

export interface ReviewPlatformThread {
  id: string;
  providerThreadId?: string | null;
  providerCommentId?: string | null;
  kind: 'review' | 'comment';
  replyToProviderCommentId?: string | null;
  filePath?: string | null;
  line?: number | null;
  resolved: boolean;
  author: string;
  body: string;
  updatedAt: string;
}

export interface ReviewPlatformPullRequestDetail extends ReviewPlatformPullRequest {
  limitations?: string[];
  body: string;
  ci: ReviewPlatformCiItem[];
  files: ReviewPlatformFile[];
  commits: ReviewPlatformCommit[];
  threads: ReviewPlatformThread[];
}

export interface ReviewPlatformPullRequestDetailPage extends ReviewPlatformPullRequestDetail {
  section: ReviewPlatformDetailSection;
  pagination: ReviewPlatformPagination;
}

export interface ReviewPlatformCiLog {
  ciItemId: string;
  log?: string | null;
  truncated: boolean;
  message?: string | null;
}

export interface ReviewPlatformCapabilities {
  canCreateReview: boolean;
  canCreatePullRequest: boolean;
  canReplyToThread: boolean;
  canResolveThread: boolean;
  canApprove: boolean;
  canRevokeApproval: boolean;
  canRequestChanges: boolean;
  canMerge: boolean;
  supportsDraftReview: boolean;
  supportedPullRequestStates?: ReviewPlatformListState[];
}

export interface ReviewPlatformPagination {
  page: number;
  perPage: number;
  total?: number | null;
  hasNext: boolean;
}

export interface ReviewPlatformWorkspaceSnapshot {
  remotes: ReviewPlatformRemote[];
  selectedRemoteId?: string | null;
  accounts: ReviewPlatformAccount[];
  repository: ReviewPlatformRepositoryRef | null;
  pullRequests: ReviewPlatformPullRequest[];
  pagination: ReviewPlatformPagination;
  capabilities: ReviewPlatformCapabilities;
  message?: string | null;
  authChallenge?: ReviewPlatformAuthChallenge | null;
}

/**
 * Identifies the repository a review-platform request runs against.
 * `workspaceId` is the authoritative identity (local vs. remote routing);
 * `repositoryPath` is only the IO operand inside that workspace.
 */
export interface ReviewRepositoryLocator {
  workspaceId: string;
  repositoryPath: string;
}

export interface ReviewPlatformWorkspaceSnapshotRequest extends ReviewRepositoryLocator {
  remoteId?: string | null;
  page?: number;
  perPage?: number;
  state?: ReviewPlatformListState;
}

export interface ReviewPlatformWorkspaceContextRequest extends ReviewRepositoryLocator {
  remoteId?: string | null;
}

export interface ReviewPlatformPullRequestDetailRequest extends ReviewRepositoryLocator {
  remoteId: string;
  pullRequestId: string;
}

export interface ReviewPlatformIssueRequest {
  platform: ReviewPlatformKind;
  host: string;
  projectPath: string;
  issueId: string;
  page?: number;
  perPage?: number;
  workspaceId?: string | null;
  repositoryPath?: string | null;
}

export interface ReviewPlatformPullRequestIdentityRequest {
  platform: ReviewPlatformKind;
  host: string;
  projectPath: string;
  pullRequestId: string;
  workspaceId?: string | null;
  repositoryPath?: string | null;
}

export interface ReviewPlatformPullRequestDetailPageRequest extends ReviewPlatformPullRequestDetailRequest {
  section: ReviewPlatformDetailSection;
  page?: number;
  perPage?: number;
}

export interface ReviewPlatformPullRequestCiLogRequest extends ReviewPlatformPullRequestDetailRequest {
  ciItemId: string;
  ciItemName: string;
}

export interface ReviewPlatformUpdateAuthTokenRequest {
  platform: ReviewPlatformKind;
  host: string;
  token: string;
}

export interface ReviewPlatformClearAuthTokenRequest {
  platform: ReviewPlatformKind;
  host: string;
}

export class ReviewPlatformAPI {
  async getWorkspaceSnapshot(
    repository: ReviewRepositoryLocator,
    remoteId?: string | null,
    page?: number,
    perPage?: number,
    state?: ReviewPlatformListState,
  ): Promise<ReviewPlatformWorkspaceSnapshot> {
    const { workspaceId, repositoryPath } = repository;
    try {
      const snapshot = await api.invoke<ReviewPlatformWorkspaceSnapshot>('review_platform_get_workspace_snapshot', {
        request: { workspaceId, repositoryPath, remoteId, page, perPage, ...(state && state !== 'all' ? { state } : {}) },
      });
      // Older hosts may ignore a new optional request field. Require their
      // advertised capability before accepting the returned page as filtered.
      if (state && state !== 'all' && !snapshot.capabilities.supportedPullRequestStates?.includes(state)) {
        throw new Error('review_platform_state_filter_unsupported');
      }
      return snapshot;
    } catch (error) {
      log.error('Failed to load review platform snapshot', { workspaceId, repositoryPath, remoteId, page, perPage, error });
      throw createTauriCommandError('review_platform_get_workspace_snapshot', error, {
        workspaceId,
        repositoryPath,
        remoteId,
        page,
        perPage,
      });
    }
  }

  async getWorkspaceContext(
    repository: ReviewRepositoryLocator,
    remoteId?: string | null,
  ): Promise<ReviewPlatformWorkspaceSnapshot> {
    const { workspaceId, repositoryPath } = repository;
    try {
      return await api.invoke('review_platform_get_workspace_context', {
        request: { workspaceId, repositoryPath, remoteId },
      });
    } catch (error) {
      log.error('Failed to load review platform workspace context', { workspaceId, repositoryPath, remoteId, error });
      throw createTauriCommandError('review_platform_get_workspace_context', error, {
        workspaceId,
        repositoryPath,
        remoteId,
      });
    }
  }

  async getPullRequestDetail(
    repository: ReviewRepositoryLocator,
    remoteId: string,
    pullRequestId: string,
  ): Promise<ReviewPlatformPullRequestDetail> {
    const { workspaceId, repositoryPath } = repository;
    try {
      return await api.invoke('review_platform_get_pull_request_detail', {
        request: { workspaceId, repositoryPath, remoteId, pullRequestId },
      });
    } catch (error) {
      log.error('Failed to load review platform pull request detail', {
        repositoryPath,
        remoteId,
        pullRequestId,
        error,
      });
      throw createTauriCommandError('review_platform_get_pull_request_detail', error, {
        repositoryPath,
        remoteId,
        pullRequestId,
      });
    }
  }

  async getPullRequestReviewTarget(
    repository: ReviewRepositoryLocator,
    remoteId: string,
    pullRequestId: string,
  ): Promise<ReviewPlatformPullRequestReviewTarget> {
    const { workspaceId, repositoryPath } = repository;
    try {
      return await api.invoke('review_platform_get_pull_request_review_target', {
        request: { workspaceId, repositoryPath, remoteId, pullRequestId },
      });
    } catch (error) {
      log.error('Failed to prepare review platform pull request target', {
        repositoryPath,
        remoteId,
        pullRequestId,
        error,
      });
      throw createTauriCommandError('review_platform_get_pull_request_review_target', error, {
        repositoryPath,
        remoteId,
        pullRequestId,
      });
    }
  }

  async getIssue(request: ReviewPlatformIssueRequest): Promise<ReviewPlatformIssueEvidence> {
    try {
      return await api.invoke('review_platform_get_issue', { request });
    } catch (error) {
      log.error('Failed to load review platform Issue evidence', {
        platform: request.platform,
        host: request.host,
        projectPath: request.projectPath,
        issueId: request.issueId,
        page: request.page,
        perPage: request.perPage,
        error,
      });
      throw createTauriCommandError('review_platform_get_issue', error, request);
    }
  }

  async getPullRequestReviewTargetByIdentity(
    request: ReviewPlatformPullRequestIdentityRequest,
  ): Promise<ReviewPlatformPullRequestReviewTarget> {
    try {
      return await api.invoke(
        'review_platform_get_pull_request_review_target_by_identity',
        { request },
      );
    } catch (error) {
      log.error('Failed to prepare review platform pull request target by identity', {
        platform: request.platform,
        host: request.host,
        projectPath: request.projectPath,
        pullRequestId: request.pullRequestId,
        error,
      });
      throw createTauriCommandError(
        'review_platform_get_pull_request_review_target_by_identity',
        error,
        request,
      );
    }
  }

  async getPullRequestDetailPage(
    request: ReviewPlatformPullRequestDetailPageRequest,
  ): Promise<ReviewPlatformPullRequestDetailPage> {
    try {
      return await api.invoke('review_platform_get_pull_request_detail_page', {
        request,
      });
    } catch (error) {
      log.error('Failed to load review platform pull request detail page', {
        repositoryPath: request.repositoryPath,
        remoteId: request.remoteId,
        pullRequestId: request.pullRequestId,
        section: request.section,
        page: request.page,
        perPage: request.perPage,
        error,
      });
      throw createTauriCommandError('review_platform_get_pull_request_detail_page', error, request);
    }
  }

  async getPullRequestCiLog(
    request: ReviewPlatformPullRequestCiLogRequest,
  ): Promise<ReviewPlatformCiLog> {
    try {
      return await api.invoke('review_platform_get_pull_request_ci_log', {
        request,
      });
    } catch (error) {
      log.error('Failed to load review platform CI log', {
        repositoryPath: request.repositoryPath,
        remoteId: request.remoteId,
        pullRequestId: request.pullRequestId,
        ciItemId: request.ciItemId,
        error,
      });
      throw createTauriCommandError('review_platform_get_pull_request_ci_log', error, request);
    }
  }

  async updateAuthToken(request: ReviewPlatformUpdateAuthTokenRequest): Promise<void> {
    try {
      await api.invoke('review_platform_update_auth_token', { request });
    } catch (error) {
      log.error('Failed to update review platform auth token', {
        platform: request.platform,
        host: request.host,
        error,
      });
      throw createTauriCommandError('review_platform_update_auth_token', error, {
        platform: request.platform,
        host: request.host,
      });
    }
  }

  async clearAuthToken(request: ReviewPlatformClearAuthTokenRequest): Promise<void> {
    try {
      await api.invoke('review_platform_clear_auth_token', { request });
    } catch (error) {
      log.error('Failed to clear review platform auth token', {
        platform: request.platform,
        host: request.host,
        error,
      });
      throw createTauriCommandError('review_platform_clear_auth_token', error, {
        platform: request.platform,
        host: request.host,
      });
    }
  }
}

export const reviewPlatformAPI = new ReviewPlatformAPI();

File: src/apps/desktop/src/api/review_platform_api.rs
//! Review platform Tauri commands.

use crate::api::app_state::AppState;
use log::error;
use openbitfun_core::service::review_platform::{
    untrusted_repository_error_message, ReviewPlatformCiLog, ReviewPlatformDetailSection,
    ReviewPlatformError, ReviewPlatformIssueEvidence, ReviewPlatformKind, ReviewPlatformListState,
    ReviewPlatformPullRequestDetail, ReviewPlatformPullRequestDetailPage,
    ReviewPlatformPullRequestReviewTarget, ReviewPlatformService, ReviewPlatformWorkspaceSnapshot,
    ReviewRepositoryLocator,
};
use serde::Deserialize;
use tauri::State;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformWorkspaceSnapshotRequest {
    /// Owning workspace ID; authoritative for local/remote Git routing.
    #[serde(default)]
    pub workspace_id: Option<String>,
    pub repository_path: String,
    pub remote_id: Option<String>,
    pub page: Option<u32>,
    pub per_page: Option<u32>,
    #[serde(default)]
    pub state: ReviewPlatformListState,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformWorkspaceContextRequest {
    #[serde(default)]
    pub workspace_id: Option<String>,
    pub repository_path: String,
    pub remote_id: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformPullRequestDetailRequest {
    #[serde(default)]
    pub workspace_id: Option<String>,
    pub repository_path: String,
    pub remote_id: String,
    pub pull_request_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformPullRequestDetailPageRequest {
    #[serde(default)]
    pub workspace_id: Option<String>,
    pub repository_path: String,
    pub remote_id: String,
    pub pull_request_id: String,
    pub section: ReviewPlatformDetailSection,
    pub page: Option<u32>,
    pub per_page: Option<u32>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformPullRequestCiLogRequest {
    #[serde(default)]
    pub workspace_id: Option<String>,
    pub repository_path: String,
    pub remote_id: String,
    pub pull_request_id: String,
    pub ci_item_id: String,
    pub ci_item_name: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformUpdateAuthTokenRequest {
    pub platform: ReviewPlatformKind,
    pub host: String,
    pub token: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformClearAuthTokenRequest {
    pub platform: ReviewPlatformKind,
    pub host: String,
}

fn review_repository(
    workspace_id: &Option<String>,
    repository_path: &str,
) -> ReviewRepositoryLocator {
    ReviewRepositoryLocator::new(workspace_id.clone(), repository_path)
}

fn optional_review_repository(
    workspace_id: &Option<String>,
    repository_path: Option<&str>,
) -> Option<ReviewRepositoryLocator> {
    repository_path.map(|path| ReviewRepositoryLocator::new(workspace_id.clone(), path))
}

#[tauri::command]
pub async fn review_platform_get_workspace_snapshot(
    _state: State<'_, AppState>,
    request: ReviewPlatformWorkspaceSnapshotRequest,
) -> Result<ReviewPlatformWorkspaceSnapshot, String> {
    ReviewPlatformService::workspace_snapshot_with_state(
        &review_repository(&request.workspace_id, &request.repository_path),
        request.remote_id.as_deref(),
        request.page,
        request.per_page,
        request.state,
    )
    .await
    .map_err(|error| {
        error!(
            "Failed to get review platform workspace snapshot: path={}, remote_id={:?}, error={}",
            request.repository_path, request.remote_id, error
        );
        review_platform_ui_error(&error)
    })
}

#[tauri::command]
pub async fn review_platform_get_workspace_context(
    _state: State<'_, AppState>,
    request: ReviewPlatformWorkspaceContextRequest,
) -> Result<ReviewPlatformWorkspaceSnapshot, String> {
    ReviewPlatformService::workspace_context(
        &review_repository(&request.workspace_id, &request.repository_path),
        request.remote_id.as_deref(),
    )
    .await
    .map_err(|error| {
        error!(
            "Failed to get review platform workspace context: path={}, remote_id={:?}, error={}",
            request.repository_path, request.remote_id, error
        );
        review_platform_ui_error(&error)
    })
}

#[tauri::command]
pub async fn review_platform_get_pull_request_detail(
    _state: State<'_, AppState>,
    request: ReviewPlatformPullRequestDetailRequest,
) -> Result<ReviewPlatformPullRequestDetail, String> {
    ReviewPlatformService::pull_request_detail(
        &review_repository(&request.workspace_id, &request.repository_path),
        &request.remote_id,
        &request.pull_request_id,
    )
    .await
    .map_err(|error| {
        error!(
            "Failed to get review platform pull request detail: path={}, remote_id={}, pull_request_id={}, error={}",
            request.repository_path,
            request.remote_id,
            request.pull_request_id,
            error
        );
        review_platform_ui_error(&error)
    })
}

#[tauri::command]
pub async fn review_platform_get_pull_request_review_target(
    _state: State<'_, AppState>,
    request: ReviewPlatformPullRequestDetailRequest,
) -> Result<ReviewPlatformPullRequestReviewTarget, String> {
    ReviewPlatformService::pull_request_review_target(
        &review_repository(&request.workspace_id, &request.repository_path),
        &request.remote_id,
        &request.pull_request_id,
    )
    .await
    .map_err(|error| {
        error!(
            "Failed to prepare review platform pull request target: path={}, remote_id={}, pull_request_id={}, error={}",
            request.repository_path,
            request.remote_id,
            request.pull_request_id,
            error
        );
        review_platform_ui_error(&error)
    })
}

#[tauri::command]
pub async fn review_platform_get_issue(
    _state: State<'_, AppState>,
    request: ReviewPlatformIssueRequest,
) -> Result<ReviewPlatformIssueEvidence, String> {
    ReviewPlatformService::issue(
        request.platform,
        &request.host,
        &request.project_path,
        &request.issue_id,
        request.page,
        request.per_page,
        optional_review_repository(&request.workspace_id, request.repository_path.as_deref())
            .as_ref(),
    )
    .await
    .map_err(|error| {
        let safe_error = safe_review_platform_error(&error);
        error!(
            "Failed to get review platform Issue: platform={:?}, host={}, project_path={}, issue_id={}, error={}",
            request.platform,
            request.host,
            request.project_path,
            request.issue_id,
            safe_error
        );
        review_platform_ui_error(&error)
    })
}

#[tauri::command]
pub async fn review_platform_get_pull_request_review_target_by_identity(
    _state: State<'_, AppState>,
    request: ReviewPlatformPullRequestIdentityRequest,
) -> Result<ReviewPlatformPullRequestReviewTarget, String> {
    ReviewPlatformService::pull_request_review_target_by_identity(
        request.platform,
        &request.host,
        &request.project_path,
        &request.pull_request_id,
        optional_review_repository(&request.workspace_id, request.repository_path.as_deref())
            .as_ref(),
    )
    .await
    .map_err(|error| {
        let safe_error = safe_review_platform_error(&error);
        error!(
            "Failed to prepare review platform pull request target by identity: platform={:?}, host={}, project_path={}, pull_request_id={}, error={}",
            request.platform,
            request.host,
            request.project_path,
            request.pull_request_id,
            safe_error
        );
        review_platform_ui_error(&error)
    })
}

fn review_platform_ui_error(error: &ReviewPlatformError) -> String {
    let code = match error {
        ReviewPlatformError::RepositoryUntrusted {
            repository_path, ..
        } => {
            return untrusted_repository_error_message(repository_path);
        }
        ReviewPlatformError::GitUnavailable => return error.to_string(),
        ReviewPlatformError::InvalidRepository(_) => "invalidRepository",
        ReviewPlatformError::RemoteNotFound(_) => "remoteNotFound",
        ReviewPlatformError::UnsupportedPlatform(_) => "unsupportedPlatform",
        ReviewPlatformError::Api(_) => "providerFailed",
        ReviewPlatformError::Http { status: 401, .. } => "authenticationRequired",
        ReviewPlatformError::Http { status: 403, .. } => "permissionDenied",
        ReviewPlatformError::Http { status: 404, .. } => "notFound",
        ReviewPlatformError::Http { .. } => "providerFailed",
        ReviewPlatformError::Network(_) => "networkFailed",
        ReviewPlatformError::Parse(_) => "invalidResponse",
        ReviewPlatformError::StaleTarget(_) => "staleTarget",
        ReviewPlatformError::EvidenceTooLarge { .. } => "evidenceTooLarge",
        ReviewPlatformError::TargetIsPullRequest { .. } => "targetIsPullRequest",
    };
    format!(
        "review_platform_error:{code}: {}",
        safe_review_platform_error(error)
    )
}

fn safe_review_platform_error(error: &ReviewPlatformError) -> String {
    match error {
        ReviewPlatformError::Http { status, .. } => format!("provider returned HTTP {status}"),
        ReviewPlatformError::Network(_) => "provider network request failed".to_string(),
        ReviewPlatformError::Parse(_) => "provider response could not be parsed".to_string(),
        ReviewPlatformError::StaleTarget(_) => {
            "provider target changed during evidence acquisition".to_string()
        }
        ReviewPlatformError::EvidenceTooLarge { .. } => {
            "provider evidence exceeded the allowed size".to_string()
        }
        ReviewPlatformError::TargetIsPullRequest { .. } => {
            "requested Issue is a pull request".to_string()
        }
        ReviewPlatformError::GitUnavailable => "Git is unavailable".to_string(),
        ReviewPlatformError::InvalidRepository(_) => "invalid repository".to_string(),
        ReviewPlatformError::RepositoryUntrusted { .. } => {
            "repository ownership is not trusted".to_string()
        }
        ReviewPlatformError::RemoteNotFound(_) => "provider remote was not found".to_string(),
        ReviewPlatformError::UnsupportedPlatform(_) => "unsupported provider".to_string(),
        ReviewPlatformError::Api(_) => "provider request was rejected".to_string(),
    }
}

#[tauri::command]
pub async fn review_platform_get_pull_request_detail_page(
    _state: State<'_, AppState>,
    request: ReviewPlatformPullRequestDetailPageRequest,
) -> Result<ReviewPlatformPullRequestDetailPage, String> {
    ReviewPlatformService::pull_request_detail_page(
        &review_repository(&request.workspace_id, &request.repository_path),
        &request.remote_id,
        &request.pull_request_id,
        request.section,
        request.page,
        request.per_page,
    )
    .await
    .map_err(|error| {
        error!(
            "Failed to get review platform pull request detail page: path={}, remote_id={}, pull_request_id={}, section={:?}, page={:?}, per_page={:?}, error={}",
            request.repository_path,
            request.remote_id,
            request.pull_request_id,
            request.section,
            request.page,
            request.per_page,
            error
        );
        review_platform_ui_error(&error)
    })
}

#[tauri::command]
pub async fn review_platform_get_pull_request_ci_log(
    _state: State<'_, AppState>,
    request: ReviewPlatformPullRequestCiLogRequest,
) -> Result<ReviewPlatformCiLog, String> {
    ReviewPlatformService::pull_request_ci_log(
        &review_repository(&request.workspace_id, &request.repository_path),
        &request.remote_id,
        &request.pull_request_id,
        &request.ci_item_id,
        &request.ci_item_name,
    )
    .await
    .map_err(|error| {
        error!(
            "Failed to get review platform CI log: path={}, remote_id={}, pull_request_id={}, ci_item_id={}, error={}",
            request.repository_path,
            request.remote_id,
            request.pull_request_id,
            request.ci_item_id,
            error
        );
        review_platform_ui_error(&error)
    })
}

#[tauri::command]
pub async fn review_platform_update_auth_token(
    _state: State<'_, AppState>,
    request: ReviewPlatformUpdateAuthTokenRequest,
) -> Result<(), String> {
    ReviewPlatformService::update_auth_token(request.platform, &request.host, &request.token)
        .await
        .map_err(|error| {
            error!(
                "Failed to update review platform auth token: platform={:?}, host={}, error={}",
                request.platform, request.host, error
            );
            review_platform_ui_error(&error)
        })
}

#[tauri::command]
pub async fn review_platform_clear_auth_token(
    _state: State<'_, AppState>,
    request: ReviewPlatformClearAuthTokenRequest,
) -> Result<(), String> {
    ReviewPlatformService::clear_auth_token(request.platform, &request.host)
        .await
        .map_err(|error| {
            error!(
                "Failed to clear review platform auth token: platform={:?}, host={}, error={}",
                request.platform, request.host, error
            );
            review_platform_ui_error(&error)
        })
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformIssueRequest {
    pub platform: ReviewPlatformKind,
    pub host: String,
    pub project_path: String,
    pub issue_id: String,
    pub page: Option<u32>,
    pub per_page: Option<u32>,
    #[serde(default)]
    pub workspace_id: Option<String>,
    pub repository_path: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformPullRequestIdentityRequest {
    pub platform: ReviewPlatformKind,
    pub host: String,
    pub project_path: String,
    pub pull_request_id: String,
    #[serde(default)]
    pub workspace_id: Option<String>,
    pub repository_path: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn review_platform_command_errors_preserve_the_repository_trust_code() {
        let error = ReviewPlatformError::RepositoryUntrusted {
            repository_path: "/srv/shared/repo".to_string(),
            detail: "fatal: detected dubious ownership".to_string(),
        };

        assert_eq!(
            review_platform_ui_error(&error),
            "git_repository_untrusted: /srv/shared/repo"
        );
    }

    #[test]
    fn review_platform_command_errors_use_stable_codes_for_other_failures() {
        let error = ReviewPlatformError::RemoteNotFound("origin".to_string());

        assert_eq!(
            review_platform_ui_error(&error),
            "review_platform_error:remoteNotFound: provider remote was not found"
        );
    }

    #[test]
    fn review_platform_request_wire_deserializes_issue_identity_fields() {
        let request: ReviewPlatformIssueRequest = serde_json::from_value(json!({
            "platform": "github",
            "host": "github.com",
            "projectPath": "example/repo",
            "issueId": "42",
            "repositoryPath": "D:/workspace/example",
            "page": 2,
            "perPage": 100
        }))
        .expect("Issue request wire should deserialize");

        assert_eq!(request.platform, ReviewPlatformKind::Github);
        assert_eq!(request.host, "github.com");
        assert_eq!(request.project_path, "example/repo");
        assert_eq!(request.issue_id, "42");
        assert_eq!(
            request.repository_path.as_deref(),
            Some("D:/workspace/example")
        );
        assert_eq!(request.page, Some(2));
        assert_eq!(request.per_page, Some(100));
    }

    #[test]
    fn review_platform_request_wire_deserializes_pull_request_identity_fields() {
        let request: ReviewPlatformPullRequestIdentityRequest = serde_json::from_value(json!({
            "platform": "gitlab",
            "host": "gitlab.com",
            "projectPath": "example/group/repo",
            "pullRequestId": "7",
            "repositoryPath": "D:/workspace/example"
        }))
        .expect("pull request identity wire should deserialize");

        assert_eq!(request.platform, ReviewPlatformKind::Gitlab);
        assert_eq!(request.host, "gitlab.com");
        assert_eq!(request.project_path, "example/group/repo");
        assert_eq!(request.pull_request_id, "7");
        assert_eq!(
            request.repository_path.as_deref(),
            Some("D:/workspace/example")
        );
    }
}
